editor/src/core/shared/element-path-tree.ts

```diff
@@ -25,57 +31,160 @@ export function elementPathTree(
-  let result: ElementPathTrees = {}
-  let currentPathParts: Array<ElementPathPart> = []
-  let priorPathString: string | null = null
-  // Loop over each and every path that needs to be added.
-  for (let elementPathIndex = 0; elementPathIndex < elementPaths.length; elementPathIndex++) {
-    const elementPath = elementPaths[elementPathIndex]
-    currentPathParts = []
-    priorPathString = null
-    // Loop over the array segments of the path that is being added.
-    for (let pathPartIndex = 0; pathPartIndex < elementPath.parts.length; pathPartIndex++) {
-      const pathPart = elementPath.parts[pathPartIndex]
-      currentPathParts[pathPartIndex] = []
-      let currentPathPart = currentPathParts[pathPartIndex]
-      // Loop over the parts of the of the path segment.
-      for (
-        let pathPartElementIndex = 0;
-        pathPartElementIndex < pathPart.length;
-        pathPartElementIndex++
-      ) {
-        const pathPartElement = pathPart[pathPartElementIndex]
-        currentPathPart.push(pathPartElement)
-        const subElementPath = EP.elementPath(currentPathParts)
-
-        // See if there's an already existing part of the tree with this path.
-        const subElementPathString = EP.toString(subElementPath)
-        const foundTree = result[subElementPathString]
-        let treeToTarget = foundTree ?? emptyElementPathTree(subElementPath, subElementPathString)
-
-        // Add this if it doesn't already exist.
-        if (foundTree == null) {
-          result[subElementPathString] = treeToTarget
-          // Handle adding this entry to its parent as a child.
-          if (priorPathString != null) {
-            const childrenOfPriorPath = result[priorPathString].children
-            childrenOfPriorPath.push(treeToTarget)
+function getMissingParentPaths(
+  paths: ElementPath[],
+  metadata: ElementInstanceMetadataMap,
+): ElementPath[] {
+  const missingParentPaths = new Set<ElementPath>()
+  for (const path of paths) {
+    let parent = EP.parentPath(path)
+    while (parent.parts.length > 0) {
+      if (metadata[EP.toString(parent)] == null) {
+        missingParentPaths.add(parent)
+      }
+      parent = EP.parentPath(parent)
+    }
+  }
+  return [...missingParentPaths.values()]
+}
+
+function getReorderedPaths(
+  original: ElementPath[],
+  metadata: ElementInstanceMetadataMap,
+  missingParents: ElementPath[],
+): ElementPath[] {
+  let elementsToReorderStack = original.filter(
+    (path) => findMaybeConditionalExpression(path, metadata) != null,
+  )
+  let paths = original.filter(
+    (path) =>
+      !elementsToReorderStack.some((other) => EP.pathsEqual(other, path)) && // omit conditionals, that will be reordered
+      !missingParents.some((parentPath) => EP.isDescendantOf(path, parentPath)), // omit elements that have a missing parent
+  )
+
+  while (elementsToReorderStack.length > 0) {
+    const conditional = elementsToReorderStack.shift()
```

Comment:
Can this be rewritten as an array reduce please to keep the amount of mutation to a minimum.

Reply:
https://github.com/concrete-utopia/utopia/pull/3811/commits/2ca47625827f7782014f6519a357661a50862c40